docs/_docs/30-fasterkv-manual-locking.md

```diff
@@ -0,0 +1,266 @@
+
+### LockTable Overview
+
+For records not found in memory, the `LockTable` is used. The semantics of `LockTable` entries are as follow. This is a conceptual view; implementation details are described in subsequent sections:
+- On a `Lock` call, if the key is not found in memory, the `LockTable` is searched for the Key.
+  - If it is not found, an entry is made in the `LockTable` with an empty `RecordInfo`.
+  - The requested `LockType` is then taken on the `RecordInfo` for that Key.
+- On an `Unlock` call, if the key is not found in memory, the `LockTable` is searched for the Key.
+  - If it is not found, a Debug.Fail() is issued.
+  - Otherwise, the requested `LockType` is unlocked. If this leaves the `RecordInfo` unlocked, its entry is deleted from the `LockTable`.
+- When a Read or RMW obtains a record from ON-DISK, it consults the `LockTable`; if the key is found, the locks are transferred to the retrieved recordInfo, and the `LockTable` entry is removed.
+- When an Upsert (without `LockOperations`) or Delete does not find a key in memory, it consults the `LockTable`, and if the key is found:
+  - it Seals the RecordInfo in the `LockTable`
+  - it performs the usual "append at tail of Log" operation
+  - it removes the entry from the `LockTable`
+- Because `LockTable` use does not verify that the key actually exists (as it does not issue a pending operation to ensure the requested key, and not a collision, is found in the on-disk portion), it is possible that keys will exist in the `LockTable` that do not in fact exist in the log. This is fine; if we do more than `Lock` them, then they will be added to the log at that time, and the locks applied to them.
+
+#### Insertion to LockTable due to Lock
+When a thread doing `Lock()` looks for a key in the LockTable and cannot find it, it must do a tentative insertion into the locktable, because it is possible that another thread CAS'd that key to the Tail of the log after the current thread had passed the hash table lookup:
+- We do not find the record in memory starting from current TailAddress, so we record that TailAddress as A.
+- Locktable does not have entry so we create a tentative entry in locktable
+- We check if key exists between current tail and A
+  - if yes we have to back off the LockTable entry creation by setting it Invalid and returning RETRY_NOW.
+    - Any thread trying an operation in the Lock Table on a Tentative record must spin until the Tentative bit is removed; this will be soon, because we are only following the hash chain back to A.
+    - Any waiting thread sees Invalid and in this case, it must also return RETRY_NOW.
+  - if no, we can set locktable entry as final by removing the Tentative bit
+    - Any waiting thread proceeds normally
+
+#### Removal from LockTable
+Here are the sequences of operations to remove records from the Lock Table:
+- Unlock
+  - If the lock count goes to 0, remove from `LockTable` conditionally on IsLocked == false.
```

Comment:
Since  only lock bits are relevant in LockTable, this is equivalent to saying `RecordInfo.word == 0`.